### PR TITLE
Generate bridge files via SSSOM

### PR DESCRIPTION
**Pull request by gouttegd**
⚠️ _This PR requires a version of ROBOT that supports pluggable commands._ ⚠️

This PR overhauls the pipeline that generates the bridges between Uberon/CL and other ontologies.

First, it removes a number of bridges that are considered no longer relevant, as discussed in #3047.

Then, it replaces the old Perl-based process to generate the remaining bridge by a 3-step process:

1) extract cross-references from Uberon/CL (which are the source of truth for almost all bridges) and from ZFA (source of truth for the CL-to-ZFA bridge) and turn them into a SSSOM set;
2) fetch externally maintained SSSOM sets for the foreign ontologies that already provide their mappings as SSSOM (currently, only FBbt – the PR prepares the way for more ontologies to follow, should they wish to do so);
3) combine all mapping sets and generate the bridges from them.

This implements the “phase 1” discussed in #3004.

Still to do in this PR:
- [x] Commit the newly generated bridge files.
- [x] Add documentation for the editors regarding the source of truth of each bridge.

### CODE REVIEW

**Comment by gouttegd** [comment]
CI check explicitly cancelled since the new pipeline won’t work anyway with the latest ODK release. An ODK image with support for ROBOT plugins is required.

**Comment by gouttegd** [comment]
Of note, the CL-to-ZFA bridge is (as expected) fully populated with the new pipeline, so this PR will fix #1813.

**Comment by gouttegd** [comment]
@matentzn @anitacaron Marking the PR as “ready for review“ since we now can use ROBOT plugins (meaning the test suite can pass).

I’d like #2928 to be finalised and merged before this one (so that if the new bridges cause any violation of the taxon constraints, we’ll know right away), but feel free to start reviewing here anytime.

File: src/ontology/bridge/cl-bridge-to-go.owl
Comment (matentzn) [inline src/ontology/bridge/cl-bridge-to-go.owl:55]: Somehow I thought the the new serialisers inject xsd:string everywhere.

**Review by matentzn — COMMENTED** [review]
I only have a little time today, so I am starting with making some small questions here and there, I will  power through by the end of this week (Thursday/Friday)

File: src/ontology/bridge/bridges.rules.m4
Comment (gouttegd) [reply src/ontology/bridge/bridges.rules.m4:1]: It’s the commonly used extension for files written in the [M4 macro processing language](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/m4.html).

File: src/ontology/bridge/cl-bridge-to-caro.obo
Comment (gouttegd) [reply src/ontology/bridge/cl-bridge-to-caro.obo:1]: Because I am not convinced it is necessary to keep them. With the old process, OBO files were just an intermediate step towards the production of the RDF/XML-formatted bridge files: the Perl script produced OBO files, which were then ROBOT-converted into RDF/XML. The new process can generate bridge files in RDF/XML (or really any other format supported by the OWL API) directly.

The updated custom Makefile contains a rule to generate OBO bridge files from the RDF/XML ones (so, basically doing the opposite of the old process: treating the RDF/XML files as intermediary towards the production of OBO files), in case someone wants/needs the OBO files, but that rule is not called by default.

I’d have no (strong) objection to always generate OBO files in addition to the RDF/XML files, but that seems an unnecessary complication to me.

File: src/ontology/bridge/bridges.dispatch
Comment (gouttegd) [reply src/ontology/bridge/bridges.dispatch:1]: I am open to any suggestion for a better extension. The extension serves no purpose anyway, apart from being (hopefully) indicative of what the file is for; the SSSOM plugin which uses that file couldn’t care less what the extension is. :)

Any line starting with `#` is ignored when parsing the file.

File: src/sparql/life-cycle-xrefs.sparql
Comment (gouttegd) [reply src/sparql/life-cycle-xrefs.sparql:1]: Agree that it’s a cleanup step that _could_ have been the object of a later separate PR, but since the reason why this bit needs cleaning up is to be found in this very PR, I think it’s best to do it here and now.

File: src/ontology/bridge/cl-bridge-to-go.owl
Comment (gouttegd) [reply src/ontology/bridge/cl-bridge-to-go.owl:17]: Those links have been unresolvable for a long time, so presumably nobody will miss them.

FYI, this is the document they used to point to: https://github.com/obophenotype/uberon/blob/master/reference/reference_0000026.md

File: src/ontology/uberon.Makefile
Comment (matentzn) [inline src/ontology/uberon.Makefile:1202]: Nice makefile pattern! cc @anitacaron

File: src/ontology/uberon.Makefile
Comment (matentzn) [inline src/ontology/uberon.Makefile:1209]: Simple and nice!

File: src/ontology/uberon.Makefile
Comment (matentzn) [inline src/ontology/uberon.Makefile:1354]: @anitacaron lets remember this pattern for the future: having the plugin path as dependency, and wget ing it the same way as a regular dependency!

File: src/ontology/uberon.Makefile
Comment (matentzn) [inline src/ontology/uberon.Makefile:1287]: I agree, but maybe double check with @cmungall whether the OBO files are needed as a release artefact. I don't think so!

File: src/ontology/uberon.Makefile
Comment (matentzn) [inline src/ontology/uberon.Makefile:1215]: Wow what a cool pattern, cc @anitacaron:

1. Evaluate the condition on makefile (not shell!) level: `ifeq ($(BRI),true)`
2. add all dependencies needed in the IF clock below
3. don't include anything in the else part

This is a great way to _really_ skip certain areas of the code you don't want to run (could be used for the DOSDP), for example.

File: src/ontology/uberon.Makefile
Comment (matentzn) [inline src/ontology/uberon.Makefile:1252]: why did you choose this pattern here? How can `if [ -f mirror/zfa.owl ]` ever be false, if it is listed in the dependencies? When people run with MIR=false?

**Review by matentzn — COMMENTED** [review]
This looks phenomenal. All my comments are for your concern only, i.e. I cannot see anything I absolutely _have_ to have changed for an approval. Feel free to resolve all comments, and re-request a review - I will simply approve it then.

File: src/ontology/bridge/cl-bridge-to-ehdaa2.owl
Comment (gouttegd) [reply src/ontology/bridge/cl-bridge-to-ehdaa2.owl:13]: > I only observe here that this is using the wrong dc namespace.

This is using the `http://purl.org/dc/elements/1.1/` namespace, *as in the original bridges*. No objection to switching to `http://purl.org/dc/terms/`, but that will require a change in the SSSOM plugin.

File: src/ontology/bridge/cl-bridge-to-emapa.owl
Comment (gouttegd) [reply src/ontology/bridge/cl-bridge-to-emapa.owl:20]: That must have been an oversight, I’ll add them back.

File: src/ontology/bridge/uberon-bridge-to-efo.owl
Comment (gouttegd) [reply src/ontology/bridge/uberon-bridge-to-efo.owl:1]: Nobody weighed in on the [inventory ticket](https://github.com/obophenotype/uberon/issues/3047) or on Slack to advocate for removing this bridge, so I left it in place. The use of the OBO URL prefix was present in the original bridge, so I didn’t change it.

No objection to removing that bridge entirely.

File: src/ontology/uberon-edit.obo
Comment (gouttegd) [reply src/ontology/uberon-edit.obo:57]: This is merely declaring Uberon’s -edit file as the source of truth for the pre-existing BFO bridge.

Whether that bridge is useful or not, it was there and again nobody advocated for its removal.

File: src/ontology/uberon.Makefile
Comment (gouttegd) [reply src/ontology/uberon.Makefile:122]: We are at [version 0.5](https://github.com/gouttegd/sssom-java/releases/tag/sssom-java-0.5.0) already and I think 0.6 is not too far away. :D

File: src/ontology/bridge/uberon-bridge-to-kupo.owl
Comment (gouttegd) [reply src/ontology/bridge/uberon-bridge-to-kupo.owl:1]: KUPO is a kidney-specific ontology (“Kidney and Urology Pathway”).

We already had a CL-to-KUPO bridge and cross-references in Uberon pointing to KUPO, so I saw no reason why we should not have a Uberon-to-KUPO bridge. I assume the absence of such a bridge had been an oversight.

File: src/ontology/uberon.Makefile
Comment (gouttegd) [reply src/ontology/uberon.Makefile:1252]: > How can if [ -f mirror/zfa.owl ] ever be false, if it is listed in the dependencies?

That’s… a very good point! I’ll have another look, but right now I can’t see how this makes sense.

File: src/ontology/uberon.Makefile
Comment (gouttegd) [reply src/ontology/uberon.Makefile:1215]: For reference, this is the idea mentioned in this ticket on the ODK: https://github.com/INCATools/ontology-development-kit/issues/912

File: src/ontology/bridge/cl-bridge-to-ehdaa2.owl
Comment (gouttegd) [reply src/ontology/bridge/cl-bridge-to-ehdaa2.owl:13]: The change [has been made](https://github.com/gouttegd/sssom-java/commit/265855629f9a87fad71b816587b81d6a6285000b) in sssom-java, so this will be automatically fixed when 1) a new sssom-java release will be available and 2) the Uberon Makefile will be updated to use that new release.

File: src/ontology/uberon.Makefile
Comment (gouttegd) [reply src/ontology/uberon.Makefile:1252]: Well, actually it does make sense.

The rule that creates the mirror uses a _shell_ conditional instead of a _Make_ conditional. That is, it basically looks like this:

```Make
mirror/zfa.owl:
    if [ $(MIR) = true ]; then ... fi
```

If we run under `MIR=false`, the mirror is not created _but_ from Make’s point of view, the rule that is supposed to create it has been executed, so the dependency is met.

So Make can trigger the execution of the bridge rule that depends on `mirror/zfa.owl` even though `mirror/zfa.owl` does not actually exist.

(This kind of situation is a big reason why I favor the use of Make conditionals, by the way.)

Comment (gouttegd) [reply src/ontology/uberon.Makefile:1252]: > I now think it’s better and cleaner to simply make the bridge pipeline dependent on both `BRI` and `MIR`.

Nope, actually we can’t do that either, that wouldn’t be enough. In order to be sure the `mirror/zfa.owl` file exists, we need `MIR` _and_ `IMP` to be true, because the rule to create the mirror is actually dependent on both variables:

```Make
mirror/zfa.owl:
    if [ $(MIR) = true ] && [ $(IMP) = true ]; then ... ; fi
```

Re-generating the bridges only when all three variables `BRI`, `MIR`, and `IMP` are true is not an option, because that would virtually guarantee that the bridges are _never_ regenerated (the release pipeline is run with `IMP=false`).

So I’ll stick to the pattern above. It’s ugly but it works, and it can’t be make cleaner until the standard ODK Makefile is updated to stop using those shell conditionals that lead Make to think a dependency has been met even though no files has been created.

File: src/ontology/uberon-edit.obo
Comment (matentzn) [reply src/ontology/uberon-edit.obo:57]: Sounds good!

File: src/ontology/uberon.Makefile
Comment (matentzn) [reply src/ontology/uberon.Makefile:1252]: It does mean though that we 

1. Add a potential point of failure (network dependency)
2. Add a potentially inconsistent state between bridge and MIR.



Another option could be to force create the mirror:

```
bridge.owl:
        make IMP=true MIR=true mirror/zfa.owl
        $(ROBOT)....
```

Just an idea.

Comment (gouttegd) [reply src/ontology/uberon.Makefile:1252]: > Another option could be to force create the mirror:
> Just an idea.

Thanks, it’s still somehow a hack but I like it better than my solution.

Longer term though, I think it’s something that should be fixed at the ODK level. I don’t see why mirrors should be not created when MIR is true but IMP is false.

**Comment by gouttegd** [comment]
@matentzn (and anyone else interested in this PR): The last changes were to fix the conflicts with the current state of the master branch and to add the documentation about how the bridges are maintained. The Pr does what I wanted it to do and is ready to be merged.

However, I am not entirely happy with something.

Currently, the bridge pipeline is _disabled_ in QC (the test suite runs with `BRI=false`). This is annoying because it means that an editor could introduce a bogus mapping that would cause some unsatisfiability issues when merging Uberon and the bridges, and such an issue would _not_ be caught by the test suite – it would only be caught at release time, which is when the bridge pipeline is enabled and the bridges are actually re-generated.

The obvious solution would be to, well, enable the bridge pipeline in QC. But this in turn would cause another problem. In addition to using the cross-references extracted from Uberon, the bridge pipeline fetches mappings from upstream sources (at least three sources, for now: the FBbt mapping set, the cross-references from CL, and the cross-references from ZFA) and uses them to produce the bridges. Doing that at QC time would mean that the test suite on any PR could fail for reasons that have nothing to do with the PR that is being tested, just because something changed in an upstream source (something that Uberon editors have no control on).

So I’d like to propose that we further refine the bridge pipeline to break it into two separate phases, to be run at different times:

1. One phase when we just fetches the remote mappings (again: from FBbt, CL, and ZFA) and _commit them to the repository_ (we are actually already doing that for the FBbt mapping set).
2. One phase when we use all the collected mappings (which by now will all be available _locally_ in the repository, without needing to fetch them from their upstream source) and the mappings locally extracted from Uberon to generate the bridges.

Phase 2 would be enabled under `BRI=true`, but phase 1 would only be enabled under `BRI=true` _and_ `IMP=true`. So only phase 2 would be enabled during a QC or release run. This would ensure that a bogus, unsat-causing mapping introduced in Uberon would be caught during QC, while preventing QC from failing because of changes happening upstream.

The only downside I can see is that in this setup, external mappings would never be _automatically_ refreshed. They would only be refreshed when an editor explicitly runs the bridge pipeline with `BRI=true` and `IMP=true` (something for which we could provide a convenient Make target, like `make refresh-bridges`). But I see that more as a _benefit_ than a _downside_, actually. It’s similar to the way we now deal with imports and it ensures that external changes are only brought into the ontology by explicit action.

So:

1. Do you agree with that idea?
2. If so, would you rather to have it implemented as part of this very PR, or do you prefer to merge this PR as it is and then refine the pipeline as proposed as part of another, distinct PR? I don’t mind either way; I have a slight preference for doing that as part of another PR as this one is already complicated enough, but I am not the one doing the review.

**Comment by matentzn** [comment]
I think your idea of separating in two sub-phases is the right one. For completeness:

We could consider `IMP` to refer to regular imports as well as imported mappings. in some sense, you could even drop the BRI=true criterion from importing external mapping sets (I am not saying we should, but it may be worth contemplating to subsume mapping syncing under import or component syncing).

In any case, I don't mind you implementing that separation here, in this very PR! Whatever works for you.

**Comment by gouttegd** [comment]
> in some sense, you could even drop the BRI=true criterion from importing external mapping sets

Yes. That’s how I actually plan to implement it. But because the mapping sets are only needed to generate the bridge, the rules to fetch the external mapping sets would in effect only be called as dependencies of the rules that generate the bridges – and those rules are themselves only used under `BRI=true`.

So basically, we would have:

```
sh run.sh make tmp/bridges BRI=true IMP=false
```

-> re-generates the bridges only from locally available resources (i.e. the mapping sets that have _previously_ been downloaded and committed to the repository)


```
sh run.sh make tmp/bridges BRI=true IMP=true
```

-> refreshes the mapping sets and re-generates the bridges

We _could_ also have something like

```
sh run.sh make tmp/mappings BRI=false [or true, doesn’t matter] IMP=true
```

to _only_ refresh all the mappings without immediately re-generating the bridges. Not sure there would be a need for that, but it would be easy to do.

> I don't mind you implementing that separation here, in this very PR! Whatever works for you.

OK, I’ll do it now and here then.

**Comment by gouttegd** [comment]
@matentzn @anitacaron This is done and ready.

Under the new system, bridges are always re-generated (both at release time, as before, and now also at QC time), but only using local versions of the externally maintained mapping sets, unless the bridge pipeline runs under `IMP=true` (which is never the case at release and QC time).

To force refreshing the bridges completely (including downloading fresh versions of the externally maintained mapping sets), a maintainer must run `refresh-bridges`, which will both refresh the mappings and immediately rebuild the bridges. Alternatively, one can also run `refresh-mappings` to refresh the mappings only and then commit the refreshed mappings – they will then be used the next time the bridge pipeline is run, e.g. at the next release).

**Comment by anitacaron** [comment]
> Alternatively, one can also run refresh-mappings to refresh the mappings only and then commit the refreshed mappings – they will then be used the next time the bridge pipeline is run, e.g. at the next release).

Can we have one command for refreshing imports and bridges? We're constantly refreshing imports before a release; we could add the bridges too.

**Comment by gouttegd** [comment]
I had considered that but I thought you would prefer to refresh the imports and the bridges separately.

You’re the one doing the releases usually, so if you prefer a single command, then sure. :)

I’ll add a `make refresh-external-resources` target that will refresh both the imports and the bridges.

File: src/ontology/mappings/cl-mappings.sssom.tsv
Comment (matentzn) [reply src/ontology/mappings/cl-mappings.sssom.tsv:1]: (same for zfa)

**Review by matentzn — APPROVED** [review]
Awesome!

File: src/ontology/mappings/cl-mappings.sssom.tsv
Comment (gouttegd) [reply src/ontology/mappings/cl-mappings.sssom.tsv:1]: I can update `sssom:xref-extract` to behave similarly in a future version.

I note that `sssom-py` injects a default license of `https://w3id.org/sssom/license/unspecified`. Is that the expected, recommended value for an unknown license? There’s no mention of it in the spec and the IRI does not resolve to anything.

For what it’s worth I disagree with `sssom-py`’s behaviour regarding the `mapping_set_id`. It injects an auto-generated ID like this: `https://w3id.org/sssom/mappings/2de2cb05-b347-42c8-bb68-48f97a8ccce1`. Sure, it is compliant with the spec that says the ID “should be IRI, ideally resolvable”, but I think that when a set does not have an idea, creating one that looks like it is resolvable even though it is certain _not_ to be resolvable is a bad idea.

Comment (gouttegd) [reply src/ontology/mappings/cl-mappings.sssom.tsv:1]: Note also that the CL and ZFA mapping sets are committed to the repository so that we can be sure they will be locally available when we need to create the bridges (without having to fetch external resources), but they are not intended to be published. It’s not up to Uberon to publish those sets, which do not “belong” to Uberon. It should be up to CL and ZFA respectively.

Comment (matentzn) [reply src/ontology/mappings/cl-mappings.sssom.tsv:1]: feel free to suggest better default values on the sssom issue tracker!

**Review by matentzn — APPROVED** [review]

**Pull request merged**
